src/pages/learn/security.mdx

```diff
@@ -0,0 +1,200 @@
+
+One of GraphQL's strengths is that clients can write expressive operations that reflect the relationships between the data exposed in the API. However, some queries may become cyclical when the fields in the selection set are deeply nested: 
+
+```graphql
+query {
+  hero {
+    name
+    friends {
+      name
+      friends {
+        name
+        friends {
+          name
+          friends {
+            name
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+Even when [the N+1 problem](/learn/performance/#the-n1-problem) has been remediated through batched requests to underlying data sources, overly nested fields may still place excessive load on server resources and impact API performance. 
+
+For this reason, it's a good idea to limit the maximum depth of fields that a single operation can have. Many GraphQL implementations expose configuration options that allow you to specify a maximum depth for a GraphQL document and return an error to the client if a request exceeds this limit before execution begins.
+
+Since nesting list fields can result in exponential increases to the amount of data returned, it's recommended to apply a separate smaller limit to how deeply lists can be nested.
+
+For cases where a client may have a legitimate use case for a deeply nested query and it's impractical to set a low blanket limit on all queries, you may need to rely on techniques such as [rate limiting](/#rate-limiting) or [query complexity analysis](#query-complexity-analysis).
+
+### Breadth and batch limiting
+
+In addition to limiting operation depth, there should also be guardrails in place to limit the number of top-level fields and field aliases included in a single operation.
+
+Consider what would happen during the execution of the following query operation:
+
+```graphql
+query {
+  viewer {
+    friends1: friends(limit: 1) { name }
+    friends2: friends(limit: 2) { name }
+    friends3: friends(limit: 3) { name }
+    # ...
+    friends100: friends(limit: 100) { name }
+  }
+}
+```
+
+Even though the overall depth of this query is shallow, the underlying data source for the API will still have to handle a large number of requests to resolve data for the aliased `hero` field.
+
+Similarly, a client may send a GraphQL document with many batched operations in a request:
+
+```graphql
+query NewHopeHero {
+  hero(episode: NEWHOPE) {
+    name
+  }
+}
+query EmpireHero {
+  hero(episode: EMPIRE) {
+    name
+  }
+}
+### ...
+query JediHero {
+  hero(episode: JEDI) {
+    name
+  }
+}
+```
+
+Depending on the client implementation, query batching can be a useful strategy for limiting the number of round trips to a server to fetch all of the required data to render a user interface. However, there should be an upper limit on the total number of queries allowed in a single batch.
+
+As with depth limiting, a GraphQL implementation may have configuration options to restrict operation breadth, field alias usage, and batching.
+
```

Comment:
Would be nice to add paragraph about recursive fragments :D

![CleanShot 2024-11-18 at 17 34 09@2x](https://github.com/user-attachments/assets/5a5bc984-1532-44e4-92a7-6e772f45e7de)

Reply:
A GraphQL API cannot have recursive fragments, as stated by the GraphQL specification:

https://spec.graphql.org/draft/#sec-Fragment-Spreads-Must-Not-Form-Cycles

Reply:
Interesting thanks!

Note: seems like it’s not always implemented properly — a CVE we found with our team.

https://github.com/graphql-rust/juniper/security/advisories/GHSA-4rx6-g5vg-5f3j

Reply:
That definitely seems like a bug in that implementation, missing one of the core validation rules is quite a major diversion from the spec - I would expect that to be opt-in.

It would be really good to get the https://github.com/graphql-cats/graphql-cats project (or one like it) up and running again so it can be checked if an implementation is actually spec-compliant. If anyone was interested in taking this on, I'm fairly sure it would likely get approval from the [community grant program](https://graphql.org/community/foundation/community-grant/).